lib/Hakyll/Core/Routes.hs

```diff
@@ -1,35 +1,42 @@
 --------------------------------------------------------------------------------
--- | Once a target is compiled, the user usually wants to save it to the disk.
--- This is where the 'Routes' type comes in; it determines where a certain
--- target should be written.
---
--- Suppose we have an item @foo\/bar.markdown@. We can render this to
--- @foo\/bar.html@ using:
---
--- > route "foo/bar.markdown" (setExtension ".html")
---
--- If we do not want to change the extension, we can use 'idRoute', the simplest
--- route available:
---
--- > route "foo/bar.markdown" idRoute
---
--- That will route @foo\/bar.markdown@ to @foo\/bar.markdown@.
---
--- Note that the extension says nothing about the content! If you set the
--- extension to @.html@, it is your own responsibility to ensure that the
--- content is indeed HTML.
---
--- Finally, some special cases:
---
--- * If there is no route for an item, this item will not be routed, so it will
---   not appear in your site directory.
---
--- * If an item matches multiple routes, the first rule will be chosen.
+{- | 'Routes' is part of the 'Hakyll.Core.Rules.Rules' processing pipeline.
+It determines if and where the compilation result of the underlying 'Hakyll.Core.Item.Item'
+being processed is written out to
+(relative to the output folder as configured in 'Hakyll.Core.Configuration.destinationDirectory').
+
+* __If there is no route for an item, the compiled item won't be written out to a file__
+and so won't appear in the output site directory.
+
+* If an item matches multiple routes, the first route will be chosen.
+
+__Examples__
+
+Suppose we have a markdown file @posts\/hakyll.md@. We can route/output its compilation result to
+@posts\/hakyll.html@ using 'setExtension':
+
+> -- file on disk: '<project-folder>/posts/hakyll.md'
+> match "posts/*" $ do
+>     route (setExtension "html") -- compilation result is written to '<output-folder>/posts/hakyll.html'
+>     compile pandocCompiler
+Hint: You can configure the output folder with 'Hakyll.Core.Configuration.providerDirectory'.
```

Comment:
Changed. Yes, the terms are intentional, i would like to describe a rule more generally as a input, transform, output pipeline. So, the terms are "input vs output" (instead of "provider vs destination"). I suppose, we could also use "source vs destination" (although the better antonym for source would be "target"). So, i can change the terms if you insist.

Reply:
I'm not sure that a pipeline is a good metaphor, because a "transformation" can have multiple "inputs" (using `load`, `loadAndApplyTemplate` etc.) and multiple "outputs" (snapshots). It also doesn't mesh well with `create`, because the "input" there is empty — how would one "transform" an empty value?

I'm also starting to doubt the idea of explaining it in this doc at all. Won't it be better to refer to the tutorial? [Part three](https://jaspervdj.be/hakyll/tutorials/03-rules-routes-compilers.html) deals with rules quite definitely,

Reply:
Ok. This metaphor (or the lack thereof) discussion is relevant for the other Rules docs PR #912; i'll keep that in mind. For this PR i've changed the wording to destination directory as you suggested.